src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java

```diff
@@ -257,12 +255,11 @@ private synchronized SchemaTransformationResult reload()
     @Override
     public SchemaTransformationResult reset(boolean local)
     {
-        return local
-               ? reload()
-               : migrationCoordinator.pullSchemaFromAnyNode()
-                                     .flatMap(mutations -> ImmediateFuture.success(applyMutations(mutations)))
-                                     .awaitThrowUncheckedOnInterrupt()
-                                     .getNow();
+        if (local)
+            return reload();
+
+        Collection<Mutation> mutations = migrationCoordinator.pullSchemaFromAnyNode().awaitThrowUncheckedOnInterrupt().getNow();
+        return applyMutations(mutations);
```

Comment:
Yeah, so in the previous scenario, `reset` was called from `Schema.resetLocalSchema`, in a thread 1. In this method, we pull the schema from some node, which is actually processed asynchronously in thread 2 and returns `Future`. Hence we use `flatMap`, `applyMutations` was called in thread 2. It ended up in a deadlock because `Schema.resetLocalSchema`, which is now synchronized was waiting for the `reset` to finish, while `applyMutations` called from thread 2 was waiting for updating `Schema` in thread 2. 

Since it is blocking anyway, we firstly fetch the schema using whichever thread, and then continue with `applyMutations` in the main thread which initiated the schema changes

Reply:
Got it. Sorry, first day after sickness and still a bit slow what's going on.